File: ha/config/custom_components/demkit/demkit/model/battery_info.py
from dataclasses import dataclass  # noqa: D100
from enum import Enum
from typing import Optional


class BatteryStatus(Enum):
    """Enum representing the status of a battery.

    Attributes:
        CHARGING (str): The battery is currently charging.
        DISCHARGING (str): The battery is currently discharging.
        IDLE (str): The battery is idle and not charging or discharging.
        ERROR (str): There is an error with the battery.

    """

    CHARGING = 0
    DISCHARGING = 1
    IDLE = 2
    ERROR = -1


@dataclass
class BatteryInfo:
    """A class to represent the information of a battery.

    Attributes:
        capacity (float): The total capacity of the battery in Wh.
        max_charge (float): The maximum charge rate of the battery in W.
        max_discharge (float): The maximum discharge rate of the battery in W.
        state_of_charge (float): The current state of charge of the battery as a percentage (0-100%).
        status (BatteryStatus): The current status of the battery.
        consumption (float): The current consumption of the battery in W.

    """

    capacity: float
    max_charge: float
    max_discharge: float
    state_of_charge: float
    target_soc: Optional[float]
    status: BatteryStatus
    consumption: float

File: hems_application/lib/model/api/battery_info.dart
import 'battery_status.dart';

/// Data wrapper class for handling backend requests.
class BatteryInfo {
  double capacity = 0;
  double consumption = 0;
  double maxCharge = 0;
  double maxDischarge = 0;
  double stateOfCharge = 0;
  BatteryStatus status = BatteryStatus.idle;
  double? targetSoc;

  BatteryInfo(
    this.capacity,
    this.consumption,
    this.maxCharge,
    this.maxDischarge,
    this.stateOfCharge,
    this.status, [
    this.targetSoc,
  ]);

  BatteryInfo.fromJson(Map<String, dynamic> json) {
    capacity = (json['capacity'] as num).toDouble();
    consumption = (json['consumption'] as num).toDouble();
    maxCharge = (json['max_charge'] as num).toDouble();
    maxDischarge = (json['max_discharge'] as num).toDouble();
    stateOfCharge = (json['state_of_charge'] as num).toDouble();

    status = switch (json['status']) {
      'Charging' => BatteryStatus.charging,
      'Discharging' => BatteryStatus.discharging,
      'Idle' => BatteryStatus.idle,
      _ => throw FormatException('Invalid battery status'),
    };

    targetSoc = (json['target_soc'] as num?)?.toDouble();
  }

  @override
  int get hashCode => Object.hash(
    capacity,
    consumption,
    maxCharge,
    maxDischarge,
    stateOfCharge,
    status,
    targetSoc,
  );

  @override
  bool operator ==(Object other) {
    return other is BatteryInfo &&
        other.capacity == capacity &&
        other.consumption == consumption &&
        other.maxCharge == maxCharge &&
        other.maxDischarge == maxDischarge &&
        other.stateOfCharge == stateOfCharge &&
        other.status == status &&
        other.targetSoc == targetSoc;
  }
}
